llm/kv_cache.go

```diff
@@ -0,0 +1,122 @@
+type GGMLModel interface {
+	KV() KV
+}
+
+// ValidKVCacheTypes contains all supported KV cache types
+var ValidKVCacheTypes = []string{"f32", "f16", "q8_0", "q4_0"}
+
+// modelSupportsFlashAttention checks if the model supports flash attention
+func modelSupportsFlashAttention(ggml GGMLModel) bool {
+	// Check if it's an embedding model - embedding models don't support flash attention
+	if _, ok := ggml.KV()[fmt.Sprintf("%s.pooling_type", ggml.KV().Architecture())]; ok {
+		return false
+	}
+
+	// Check head counts match and are non-zero
+	headCountK := ggml.KV().EmbeddingHeadCountK()
+	headCountV := ggml.KV().EmbeddingHeadCountV()
+	return headCountK != 0 && headCountV != 0 && headCountK == headCountV
+}
+
+// checkFlashAttentionSupport returns whether flash attention should be enabled and a reason if it's not
+func checkFlashAttentionSupport(ggml GGMLModel, gpus discover.GpuInfoList) (bool, string) {
+	flashAttnRequested := envconfig.FlashAttention()
+	if !flashAttnRequested {
+		return false, "not requested"
+	}
+
+	if !gpus.SupportsFlashAttention() {
+		return false, "not supported by GPU"
+	}
+
+	if !modelSupportsFlashAttention(ggml) {
+		return false, "not supported by model"
+	}
+
+	return true, ""
+}
+
+// kVCacheQuantization determines the appropriate KV cache quantization type based on
+// model characteristics, GPU capabilities, and environment settings
+func kVCacheQuantization(ggml GGMLModel, gpus discover.GpuInfoList) string {
+	kvct := envconfig.KvCacheType()
+	if kvct == "" {
+		return "f16" // default value
+	}
+
+	// First check if the requested type is valid
+	if !slices.Contains(ValidKVCacheTypes, kvct) {
+		slog.Warn("unsupported cache type, defaulting to f16", "type", kvct)
+		return "f16"
+	}
+
+	// Check if it's an embedding model first
+	if _, isEmbedding := ggml.KV()[fmt.Sprintf("%s.pooling_type", ggml.KV().Architecture())]; isEmbedding {
+		if kvct == "f32" {
+			return "f32"
+		}
+		if kvct != "f16" {
+			slog.Warn("only f16 and f32 cache types are supported for embedding models, defaulting to f16",
+				"requested_type", kvct)
+		}
+		return "f16"
+	}
+
+	// Check flash attention support
+	flashAttnEnabled, _ := checkFlashAttentionSupport(ggml, gpus)
+	if !flashAttnEnabled {
+		if kvct != "f16" && kvct != "f32" {
+			slog.Info("quantized KV cache requires flash attention, defaulting to f16",
+				"requested_type", kvct)
+		}
+		return "f16"
+	}
+
+	// Flash attention is enabled and type is valid, use requested type
+	return kvct
+}
+
+// estimateKvCacheSize determines the total memory required for KV cache based on the quantization type
+func estimateKvCacheSize(cacheType string, numCtx uint64, ggml *GGML) uint64 {
+	var bytesPerElement float64
+	isEmbeddingModel := !modelSupportsFlashAttention(ggml)
+
+	// For embedding models, automatically convert to f16 unless f32 is explicitly requested
+	if isEmbeddingModel && cacheType != "f32" {
+		cacheType = "f16"
+	}
```

Comment:
To clean this up, the caller should check (and adjust) the cacheType ahead of time before passing to this function – that way it's only used for estimation

Reply:
I'm not sure I like this as it would mean that any time the function is used you're:

1. Assuming the kvct has already been tested.
2. Potentially duplicating the testing code.